### PR TITLE
`r\hpc_cache`: Add support for encryption key

### PR DESCRIPTION
**Pull request by myc2h6o**
- This feature internally uses Disk Encryption Set so validations for Key Vault are same as it here https://github.com/hashicorp/terraform-provider-azurerm/blob/82de9ae14d58e97d70664acda90b2036852f92df/internal/services/compute/disk_encryption_set_resource.go#L113
- When updating `key_vault_key_id`, `auto_key_rotation_enabled` will be set to `false` at backend, have confirmed with service team that this is by design, so added validation to enforce `auto_key_rotation_enabled` to be `false` when changing `key_vault_key_id`

### CODE REVIEW

File: internal/services/hpccache/hpc_cache_resource.go [hunk 1]
```diff
@@ -923,6 +1104,46 @@ func resourceHPCCacheSchema() map[string]*pluginsdk.Schema {
...
+						Type:     pluginsdk.TypeSet,
+						Optional: true,
+						Elem: &pluginsdk.Schema{
+							Type:         pluginsdk.TypeString,
+							ValidateFunc: msivalidate.UserAssignedIdentityID,
+						},
+					},
+				},
+			},
+		},
```
Comment (tombuildsstuff) [inline internal/services/hpccache/hpc_cache_resource.go:1132]: this needs to be using Common Schema: https://github.com/hashicorp/go-azure-helpers/blob/main/resourcemanager/commonschema/identity_user.go

File: website/docs/r/hpc_cache.html.markdown [hunk 2]
```diff
@@ -83,7 +83,15 @@ The following arguments are supported:
 * `directory_ldap` - (Optional) A `directory_ldap` block as defined below.
 
 ~> **Note:** Only one of `directory_active_directory`, `directory_flat_file` and `directory_ldap` can be set.
- 
+
+* `identity` - (Optional) An `identity` block as defined below.
+
+* `key_vault_key_id` - (Optional) Specifies the URL to a Key Vault Key.
```
Comment (tombuildsstuff) [inline website/docs/r/hpc_cache.html.markdown:89]: is this the Resource Manager or Data Plane Key ID?

File: website/docs/r/hpc_cache.html.markdown [hunk 3]
```diff
@@ -83,7 +83,15 @@ The following arguments are supported:
...
+
+~> **NOTE:** `auto_key_rotation_enabled` must be set to `false` when updating `key_vault_key_id`.
```
Comment (tombuildsstuff) [inline website/docs/r/hpc_cache.html.markdown:91]: that'd be a provider behaviour we're missing, to disable this, then re-enable it during the update?

File: website/docs/r/hpc_cache.html.markdown [hunk 4]
```diff
@@ -83,7 +83,15 @@ The following arguments are supported:
 * `directory_ldap` - (Optional) A `directory_ldap` block as defined below.
 
 ~> **Note:** Only one of `directory_active_directory`, `directory_flat_file` and `directory_ldap` can be set.
- 
+
+* `identity` - (Optional) An `identity` block as defined below.
```
Comment (tombuildsstuff) [inline website/docs/r/hpc_cache.html.markdown:87]: this is ForceNew

File: website/docs/r/hpc_cache.html.markdown [hunk 5]
```diff
@@ -172,6 +180,14 @@ A `dns` block contains the following:
 
 * `search_domain` - (Optional) The DNS search domain for the HPC Cache.
 
+---
+
+An `identity` block supports the following:
+
+* `type` - (Required) Specifies the type of Managed Service Identity that should be configured on this HPC Cache. Possible value is `UserAssigned`.
+
+* `identity_ids` - (Optional) Specifies a list of User Assigned Managed Identity IDs to be assigned to this HPC Cache.
```
Comment (tombuildsstuff) [inline website/docs/r/hpc_cache.html.markdown:189]: there's common phrasing for these now, can we pull these from another resource

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @myc2h6o 

Thanks for this PR - I've taken a look through and left a few comments inline, if we can fix those comments up then we should be able to take another look.

Thanks!

File: internal/services/hpccache/hpc_cache_resource.go (hunk 1 above)
Comment (myc2h6o) [reply internal/services/hpccache/hpc_cache_resource.go]: have updated to `commonschema.UserAssignedIdentityOptionalForceNew()`

File: website/docs/r/hpc_cache.html.markdown (hunk 2 above)
Comment (myc2h6o) [reply website/docs/r/hpc_cache.html.markdown]: this is Resource Managed ID, which is `azurerm_key_vault_key.id`, have updated the description to be clear

File: website/docs/r/hpc_cache.html.markdown (hunk 3 above)
Comment (myc2h6o) [reply website/docs/r/hpc_cache.html.markdown]: Yes make sense, I've updated the code to trigger another update when the key is updated

File: website/docs/r/hpc_cache.html.markdown (hunk 4 above)
Comment (myc2h6o) [reply website/docs/r/hpc_cache.html.markdown]: Added the forceNew description

File: website/docs/r/hpc_cache.html.markdown (hunk 5 above)
Comment (myc2h6o) [reply website/docs/r/hpc_cache.html.markdown]: Have updated to use the common phrasing

**Comment by myc2h6o** [comment]
Hi @tombuildsstuff thanks for reviewing the change! I've resolved the comment, could you please take another look?

**Comment by myc2h6o** [comment]
Updated test result: (Failed test `TestAccHPCCache_directoryLDAP` fails intermittently on main, to be fixed by #17060)

image

**Review by katbyte — CHANGES_REQUESTED** [review]

File: internal/services/hpccache/hpc_cache_resource.go [hunk 6]
```diff
@@ -251,6 +319,28 @@ func resourceHPCCacheRead(d *pluginsdk.ResourceData, meta interface{}) error {
...
+		if props.KeyEncryptionKey != nil && props.KeyEncryptionKey.KeyURL != nil {
+			keyVaultKeyId = *props.KeyEncryptionKey.KeyURL
+		}
+
+		if props.RotationToLatestKeyVersionEnabled != nil {
+			autoKeyRotationEnabled = *props.RotationToLatestKeyVersionEnabled
+		}
+	}
+	d.Set("key_vault_key_id", keyVaultKeyId)
+	d.Set("auto_key_rotation_enabled", autoKeyRotationEnabled)
```
[the remark this answers is not in the document]
Comment (myc2h6o) [reply internal/services/hpccache/hpc_cache_resource.go]: @katbyte I've renamed the property name.

**Comment by myc2h6o** [comment]
Updated test result
image

**Review by katbyte — APPROVED** [review]
thanks @myc2h6o - LGTM 🌻

**Pull request merged**